File: bluefire/native_tool_installations.py
"""Strict identity records for reviewed, preinstalled native tools.

This module parses and hashes setup metadata only.  It deliberately does not
inspect the filesystem, establish readiness, or dispatch an executable.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from types import MappingProxyType
from typing import Any, Mapping, cast

from .contracts import ContractError
from .util import content_hash, json_clone

SCHEMA = "bluefire.native-tool-installation.v1"
_FIELDS = frozenset(
    {
        "schema_version",
        "adapter_id",
        "adapter_version",
        "adapter_contract_digest",
        "tool_id",
        "tool_version",
        "platform",
        "architecture",
        "content_sha256",
        "size_bytes",
        "installation_location",
    }
)
_STABLE_ID = re.compile(r"^[a-z][a-z0-9]*(?:[._-][a-z0-9]+)*\.v[1-9][0-9]*$")
_VERSION = re.compile(r"^[0-9]+\.[0-9]+\.[0-9]+$", re.ASCII)
_TOOL_VERSION = re.compile(r"^[A-Za-z0-9][A-Za-z0-9.-]{0,63}$", re.ASCII)
_DIGEST = re.compile(r"^sha256:[0-9a-f]{64}$")
_ARCHITECTURES = frozenset({"x86_64", "aarch64"})
_MAX_SIZE = 128 * 1024 * 1024
CANDIDATE_SCHEMA = "bluefire.native-tool-candidate.v1"


def _text(value: Any, name: str, pattern: re.Pattern[str]) -> str:
    if (
        not isinstance(value, str)
        or len(value) > 256
        or pattern.fullmatch(value) is None
        or any(ord(char) < 32 for char in value)
    ):
        raise ContractError(f"native tool installation {name} is invalid")
    return value


def _location(value: Any) -> str:
    if not isinstance(value, str) or not 1 <= len(value) <= 4096:
        raise ContractError("native tool installation location is invalid")
    if not value.startswith("/") or value.startswith("//"):
        raise ContractError("native tool installation location must be absolute POSIX")
    if len(value) > 1 and value.endswith("/"):
        raise ContractError("native tool installation location is not canonical")
    if "\\" in value or any(
        ord(char) < 32 or 127 <= ord(char) < 160 or 0xD800 <= ord(char) <= 0xDFFF for char in value
    ):
        raise ContractError("native tool installation location contains unsafe characters")
    components = value.split("/")[1:]
    if not components or any(component in {"", ".", ".."} for component in components):
        raise ContractError("native tool installation location is not canonical")
    return value


def canonical_native_tool_candidate(value: Any) -> dict[str, str]:
    """Validate read-only setup inputs without claiming installed identity."""
    if not isinstance(value, Mapping) or set(value) != {
        "schema_version",
        "action_id",
        "installation_location",
        "tool_version",
    }:
        raise ContractError("native tool candidate has an unsupported shape")
    if value["schema_version"] != CANDIDATE_SCHEMA:
        raise ContractError("native tool candidate schema is unsupported")
    return {
        "schema_version": CANDIDATE_SCHEMA,
        "action_id": _text(value["action_id"], "action_id", _STABLE_ID),
        "installation_location": _location(value["installation_location"]),
        "tool_version": _text(value["tool_version"], "tool_version", _TOOL_VERSION),
    }


@dataclass(frozen=True, slots=True, repr=False)
class NativeToolInstallation:
    """Immutable setup identity; trust is supplied by the caller's setup path."""

    _canonical: Mapping[str, Any]

    @classmethod
    def from_mapping(cls, value: Any) -> "NativeToolInstallation":
        if not isinstance(value, Mapping) or set(value) != _FIELDS:
            raise ContractError("native tool installation must contain exactly its declared fields")
        if value["schema_version"] != SCHEMA:
            raise ContractError("native tool installation schema is unsupported")
        data = {
            "schema_version": SCHEMA,
            "adapter_id": _text(value["adapter_id"], "adapter_id", _STABLE_ID),
            "adapter_version": _text(value["adapter_version"], "adapter_version", _VERSION),
            "adapter_contract_digest": _text(
                value["adapter_contract_digest"], "adapter_contract_digest", _DIGEST
            ),
            "tool_id": _text(value["tool_id"], "tool_id", _STABLE_ID),
            "tool_version": _text(value["tool_version"], "tool_version", _TOOL_VERSION),
            "platform": value["platform"],
            "architecture": value["architecture"],
            "content_sha256": _text(value["content_sha256"], "content_sha256", _DIGEST),
            "size_bytes": value["size_bytes"],
            "installation_location": _location(value["installation_location"]),
        }
        if data["platform"] != "linux":
            raise ContractError("native tool installation platform is unsupported")
        if (
            not isinstance(data["platform"], str)
            or not isinstance(data["architecture"], str)
            or data["architecture"] not in _ARCHITECTURES
        ):
            raise ContractError("native tool installation architecture is unsupported")
        if (
            isinstance(data["size_bytes"], bool)
            or not isinstance(data["size_bytes"], int)
            or not 1 <= data["size_bytes"] <= _MAX_SIZE
        ):
            raise ContractError("native tool installation size_bytes is invalid")
        return cls(MappingProxyType(json_clone(data)))

    def to_dict(self) -> dict[str, Any]:
        return cast(dict[str, Any], json_clone(dict(self._canonical)))

    @property
    def digest(self) -> str:
        return content_hash(dict(self._canonical))

    def check_binding(
        self,
        *,
        expected_adapter_id: str,
        expected_adapter_version: str,
        expected_adapter_contract_digest: str,
        expected_tool_id: str,
        expected_platform: str,
        expected_architecture: str,
    ) -> None:
        expected = {
            "adapter_id": expected_adapter_id,
            "adapter_version": expected_adapter_version,
            "adapter_contract_digest": expected_adapter_contract_digest,
            "tool_id": expected_tool_id,
            "platform": expected_platform,
            "architecture": expected_architecture,
        }
        for field, value in expected.items():
            if self._canonical[field] != value:
                raise ContractError(f"native tool installation {field} does not match binding")


__all__ = ["NativeToolInstallation", "SCHEMA"]


def canonical_native_tool_installations(
    value: Any, *, platform: str, allowed_actions: Any
) -> list[dict[str, Any]]:
    """Seal finite setup records; compiled-method admission remains runner-owned."""
    if not isinstance(value, (list, tuple)) or len(value) > 16:
        raise ContractError("native tool installations must be a list of at most 16 records")
    if not value:
        return []
    if not isinstance(allowed_actions, (list, tuple)) or any(
        not isinstance(action, str) for action in allowed_actions
    ):
        raise ContractError("native tool installations require explicit allowed actions")
    records = [NativeToolInstallation.from_mapping(item).to_dict() for item in value]
    identities = set()
    for record in records:
        if record["platform"] != platform or record["adapter_id"] not in allowed_actions:
            raise ContractError("native tool installation is outside its runner profile")
        if record["adapter_id"] in identities:
            raise ContractError("native tool installation adapter is duplicated")
        identities.add(record["adapter_id"])
    return sorted(records, key=lambda record: str(record["adapter_id"]))

File: runner/src/native_tool_installations.rs
//! Strict identity metadata for reviewed, preinstalled native tools.
//!
//! This module validates and hashes setup metadata only.  It does not inspect
//! the filesystem, establish readiness, or dispatch an executable.

use serde::{Deserialize, Serialize};

use crate::canonical::canonical_hash;

pub const SCHEMA: &str = "bluefire.native-tool-installation.v1";
const MAX_SIZE_BYTES: u64 = 128 * 1024 * 1024;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub struct NativeToolInstallation {
    pub schema_version: String,
    pub adapter_id: String,
    pub adapter_version: String,
    pub adapter_contract_digest: String,
    pub tool_id: String,
    pub tool_version: String,
    pub platform: String,
    pub architecture: String,
    pub content_sha256: String,
    pub size_bytes: u64,
    pub installation_location: String,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct NativeToolBinding {
    pub adapter_id: &'static str,
    pub adapter_version: &'static str,
    pub adapter_contract_digest: &'static str,
    pub tool_id: &'static str,
}

impl NativeToolInstallation {
    pub fn validate(&self) -> Result<(), String> {
        if self.schema_version != SCHEMA {
            return Err("native tool installation schema is unsupported".into());
        }
        stable_id(&self.adapter_id, "adapter_id")?;
        version(&self.adapter_version, "adapter_version")?;
        digest(&self.adapter_contract_digest, "adapter_contract_digest")?;
        stable_id(&self.tool_id, "tool_id")?;
        tool_version(&self.tool_version)?;
        if self.platform != "linux" {
            return Err("native tool installation platform is unsupported".into());
        }
        if !matches!(self.architecture.as_str(), "x86_64" | "aarch64") {
            return Err("native tool installation architecture is unsupported".into());
        }
        digest(&self.content_sha256, "content_sha256")?;
        if !(1..=MAX_SIZE_BYTES).contains(&self.size_bytes) {
            return Err("native tool installation size_bytes is invalid".into());
        }
        location(&self.installation_location)?;
        Ok(())
    }

    pub fn digest(&self) -> String {
        canonical_hash(&serde_json::to_value(self).expect("native tool serialization cannot fail"))
    }
}

impl NativeToolBinding {
    pub fn check_binding(
        &self,
        installation: &NativeToolInstallation,
        platform: &str,
        architecture: &str,
    ) -> Result<(), String> {
        installation.validate()?;
        let expected = [
            (
                "adapter_id",
                self.adapter_id,
                installation.adapter_id.as_str(),
            ),
            (
                "adapter_version",
                self.adapter_version,
                installation.adapter_version.as_str(),
            ),
            (
                "adapter_contract_digest",
                self.adapter_contract_digest,
                installation.adapter_contract_digest.as_str(),
            ),
            ("tool_id", self.tool_id, installation.tool_id.as_str()),
            ("platform", platform, installation.platform.as_str()),
            (
                "architecture",
                architecture,
                installation.architecture.as_str(),
            ),
        ];
        for (field, expected, actual) in expected {
            if expected != actual {
                return Err(format!(
                    "native tool installation {field} does not match binding"
                ));
            }
        }
        Ok(())
    }
}

fn text(value: &str, max_len: usize) -> bool {
    !value.is_empty() && value.chars().count() <= max_len && !value.chars().any(|c| c < '\u{20}')
}

fn stable_id(value: &str, name: &str) -> Result<(), String> {
    if !text(value, 256) {
        return Err(format!("native tool installation {name} is invalid"));
    }
    let (base, version) = value
        .rsplit_once(".v")
        .ok_or_else(|| format!("native tool installation {name} is invalid"))?;
    if version.is_empty()
        || !version.chars().all(|c| c.is_ascii_digit())
        || version.starts_with('0')
        || base.is_empty()
        || !base.chars().next().is_some_and(|c| c.is_ascii_lowercase())
    {
        return Err(format!("native tool installation {name} is invalid"));
    }
    let mut segment_start = true;
    for c in base.chars() {
        if matches!(c, '.' | '_' | '-') {
            if segment_start {
                return Err(format!("native tool installation {name} is invalid"));
            }
            segment_start = true;
        } else if c.is_ascii_lowercase() || c.is_ascii_digit() {
            segment_start = false;
        } else {
            return Err(format!("native tool installation {name} is invalid"));
        }
    }
    if segment_start {
        return Err(format!("native tool installation {name} is invalid"));
    }
    Ok(())
}

fn version(value: &str, name: &str) -> Result<(), String> {
    let valid = text(value, 256)
        && value.split('.').count() == 3
        && value
            .split('.')
            .all(|part| !part.is_empty() && part.chars().all(|c| c.is_ascii_digit()));
    if valid {
        Ok(())
    } else {
        Err(format!("native tool installation {name} is invalid"))
    }
}

fn tool_version(value: &str) -> Result<(), String> {
    let valid = text(value, 64)
        && value
            .chars()
            .next()
            .is_some_and(|c| c.is_ascii_alphanumeric())
        && value
            .chars()
            .all(|c| c.is_ascii_alphanumeric() || matches!(c, '.' | '-'));
    if valid {
        Ok(())
    } else {
        Err("native tool installation tool_version is invalid".into())
    }
}

fn digest(value: &str, name: &str) -> Result<(), String> {
    let valid = value.len() == 71
        && value.starts_with("sha256:")
        && value[7..]
            .chars()
            .all(|c| c.is_ascii_hexdigit() && !c.is_ascii_uppercase());
    if valid {
        Ok(())
    } else {
        Err(format!("native tool installation {name} is invalid"))
    }
}

fn location(value: &str) -> Result<(), String> {
    let invalid = !text(value, 4096)
        || !value.starts_with('/')
        || value.starts_with("//")
        || (value.len() > 1 && value.ends_with('/'))
        || value.contains('\\')
        || value
            .chars()
            .any(|c| c == '\u{7f}' || ('\u{80}'..='\u{9f}').contains(&c));
    if invalid
        || value
            .split('/')
            .skip(1)
            .any(|part| part.is_empty() || part == "." || part == "..")
    {
        Err("native tool installation location is not canonical".into())
    } else {
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn installation() -> NativeToolInstallation {
        NativeToolInstallation {
            schema_version: SCHEMA.into(),
            adapter_id: "sandbox.permission.chmod.v1".into(),
            adapter_version: "1.0.0".into(),
            adapter_contract_digest: format!("sha256:{}", "a".repeat(64)),
            tool_id: "gnu.coreutils.chmod.v1".into(),
            tool_version: "9.5".into(),
            platform: "linux".into(),
            architecture: "x86_64".into(),
            content_sha256: format!("sha256:{}", "b".repeat(64)),
            size_bytes: 1234,
            installation_location: "/usr/bin/chmod".into(),
        }
    }

    #[test]
    fn serde_is_strict_and_digest_is_canonical() {
        let value = installation();
        let encoded = serde_json::to_value(&value).unwrap();
        assert_eq!(value, serde_json::from_value(encoded.clone()).unwrap());
        assert!(serde_json::from_value::<NativeToolInstallation>(json!({
            "schema_version": SCHEMA,
            "adapter_id": "adapter.example.v1",
            "adapter_version": "1.2.3",
            "adapter_contract_digest": format!("sha256:{}", "a".repeat(64)),
            "tool_id": "tool.example.v1",
            "tool_version": "2026.09-linux",
            "platform": "linux",
            "architecture": "x86_64",
            "content_sha256": format!("sha256:{}", "b".repeat(64)),
            "size_bytes": 1,
            "installation_location": "/opt/bluefire/tool",
            "extra": true
        }))
        .is_err());
        assert_eq!(value.digest(), canonical_hash(&encoded));
        assert_eq!(
            value.digest(),
            "sha256:fd6a0ff27b30372ab441730505a2b3c3ae03e2c923632232ee832ab5021c16c8"
        );
    }

    #[test]
    fn validation_rejects_unsafe_identity_and_location() {
        let mut value = installation();
        value.adapter_id = "Adapter.example.v1".into();
        assert!(value.validate().is_err());
        value = installation();
        value.installation_location = "/opt/../tool".into();
        assert!(value.validate().is_err());
        value = installation();
        value.size_bytes = MAX_SIZE_BYTES + 1;
        assert!(value.validate().is_err());
    }

    #[test]
    fn binding_requires_exact_identity_and_target() {
        let value = installation();
        let binding = NativeToolBinding {
            adapter_id: "sandbox.permission.chmod.v1",
            adapter_version: "1.0.0",
            adapter_contract_digest:
                "sha256:aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
            tool_id: "gnu.coreutils.chmod.v1",
        };
        assert!(binding.check_binding(&value, "linux", "x86_64").is_ok());
        assert!(binding.check_binding(&value, "linux", "aarch64").is_err());
    }
}
